### ISSUE TITLE
Give more love to registered models

### ISSUE THREAD

**Issue by hugary1995** [applied-material-modeling/neml2#55]
Model `A` may be registered in Model `B`, so that during the evaluation of Model `B`, Model `A` can be _called_.

This is a good mechanism, and currently has three use cases: 
- associative flow rules where the yield function needs to be registered
- implicit time integration where a rate model needs to be registered which will be used to compute the residual
- implicit update where an implicit model needs to be registered and "solved" during its evaluation

The first two use cases are similar, as we just need to evaluate the registered model with the current input.
The third use case requires one additional step, which is to cache the current input, and evaluate the registered model with "frozen" forces.

To summarize, 
1. The first issue is different models may have different ways of evaluating a registered model.
2. The second issue is the same model may be registered by more than one model, in which case there might be repeated (redundant) evaluations of the same model with the same input.

To address the first issue, we just need to provide a virtual method, e.g., `prepare_value`, to be called before `set_value`.
The second issue is bit more tricky, as we not only need to cache the outputs of a registered model, but also be able to tell whether the cached output is suitable to be reused given a new input.

**Comment by hugary1995** [comment]
This is handled by 2nd order chain rule from #74
